src/RequestHandle.cpp

```diff
@@ -480,4 +538,52 @@ auto curl_write_data(
     return data_length;
 }
 
+auto curl_write_partial_data(
+    void* buffer,
+    size_t size,
+    size_t nitems,
+    void* user_ptr
+) -> size_t
+{
+    auto* raw_request_ptr = static_cast<RequestHandle*>(user_ptr);
+
+    size_t data_length = size * nitems;
+
+    auto bytes_left_to_write_optional = raw_request_ptr->getRemainingDownloadBytesOptional();
+
+    if (bytes_left_to_write_optional) {
+        auto bytes_left_to_write = bytes_left_to_write_optional.value();
+        /**
+         * If the max number of bytes left to write is less than the
+         * data_length of the current frame, just set that as the data_length
+         * when appending to the request's data string.
+         * This will also cause the download to stop, since CURL aborts
+         * when the length returned from the write callback does match the
+         * length (size * nitems) passed in.
+         **/
+        if (bytes_left_to_write < data_length) {
+            data_length = bytes_left_to_write;
+        }
+
+        raw_request_ptr->m_response_data.append(
+            static_cast<const char*>(buffer),
+            data_length
+        );
+
+        raw_request_ptr->m_bytes_written += data_length;
+    }
+    else
+    {
+        /**
+         * If we get here, we have somehow downloaded more bytes than max bytes.
+         * Therefore, set data_length to a value not expected by CURL to cause
+         * the download to abort. The status will be set to indicate the wrong
+         * number of bytes were downloaded.
+         */
+        ++data_length;
```

Comment:
this comment seems like the wrong approach, if you HAVE downloaded too many just chop them off and succeed.  realistically you shouldn't have this happen if done correctly but I don't see why you would fail if you did manage to download the requested number of bytes

Reply:
I've updated the code and comment -- I think this is closer to what you had in mind.